robocup_ws/src/Planner/src/player_controller.py

```diff
@@ -0,0 +1,128 @@
+from game_interfaces.msg import Position
+from Planner.src.brachistochrone import cycloid
+from BasicCommonActions.go_to_action import go_to_fast
+import math
+
+
+class PlayerController:  #(Robot)
+    def __init__(self, player_id):
+        self.read = True
+        self.player = player_id
+        self.points_to_visit = []
+        self.current_goal = 0
+        self.goal_threshold = 0.1
+
+    def has_ball(self, game_info):
+        team = game_info[0]
+        player_id = game_info[2]
+        position = team.players_positions_efcs[player_id]
+        ball_pos = team.ball_pos_efcs
+        d = np.hypot(ball_pos.x - position.x, ball_pos.y - position.y)
+        if (d <= self.goal_threshold):
+            return True
+        return False
+
+    def if_able_score(self, game_info):
+        pass
+
+    def check_for_pass(self, game_info):
+        team = game_info[0]
+        player_id = game_info[2]
+        positions = team.players_positions_efcs
+        position = [[position.x, position.y] for position in positions]
+        main_player = position[player_id]
+        position[player_id] = [np.inf, np.inf]
+        position = np.array(position).astype(float)
+        player_pos = np.array([main_player[0], main_player[1]]).astype(float)
+        idx = np.argmin(np.linalg.norm(position - player_pos, axis=1))
+        return True, idx
+
+    def pass_ball(self, game_info, candidate):
+        team = game_info[0]
+        main_player = team.players_positions_efcs[game_info[2]]
+        candidate_pos = team.players_positions_efcs[candidate]
+        position = np.array([candidate_pos.x,candidate_pos.y])
+        ball = team.ball_pos_efcs
+        ball_pos = np.array([ball.x, ball.y])
+        relative_vector = position - ball_pos
+        unit_vector = relative_vector/np.linalg.norm(relative_vector)
+        vantage_point = ball_pos - 0.1*unit_vector
+        kick = 1
+        if(np.linalg.norm(np.array([main_player.x, main_player.y]).astype(float) - vantage_point) <= 0.1):
+            print("VANTAGE: " + str(np.linalg.norm(np.array([main_player.x, main_player.y]).astype(float) - vantage_point)))
+            return PlayerCommand(0,0,kick)
+        lv, rv = go_to_fast(main_player,Position(vantage_point[0],vantage_point[1],0))
+        return PlayerCommand(lv,rv,0)
+
+
+    def closest_to_ball(self,game_info):
+        team = game_info[0]
+        player_id = game_info[2]
+        positions = team.players_positions_efcs
+        position = [[position.x,position.y] for position in positions]
+        position = np.array(position).astype(float)
+        ball_pos = np.array([team.ball_pos_efcs.x,team.ball_pos_efcs.y]).astype(float)
+        idx = np.argmin(np.linalg.norm(position-ball_pos, axis=1))
+        print(idx)
+        return idx
+
+    def go_to_strategic_point(self,game_info):
+        return PlayerCommand(0,0,0) #TODO
+
+    def intercept(self,game_info):
+        return self.go_to_ball(game_info)
```

Comment:
when you calculate https://github.com/szgula/UofG_Robotics_TDP/blob/a7fddf38f5438ca2089b1e7bdb62ef6a77113c7a/robocup_ws/src/Planner/src/BasicCommonActions/plan_supporting_functions.py#L100
it returns the player which is closest to ball and when it can get to the intersection point

having the player and time

https://github.com/szgula/UofG_Robotics_TDP/blob/a7fddf38f5438ca2089b1e7bdb62ef6a77113c7a/robocup_ws/src/Planner/src/BasicCommonActions/plan_supporting_functions.py#L54

will give you point where robot should go to capture the ball